Only start nomination for next block when half have signed the last block

hewison-chris opened the pull request:
This should help ensure we get more signatures on blocks as there will be more chance to catch up on signing before the next block is created. This should also help prevent too many empty blocks in a row when block creation is delayed for some reason.

hewison-chris committed:
Sign a fetched block if we were not offline during nomination

hewison-chris committed:
Nominate after more than half sign the previous block

Enforcing more signatures on last block before progressing with the
next block.

hewison-chris committed:
Increase txs from faucet for ci Integration test

Quite often 3 empty blocks are generated which then causes the test to
fail as it checks for not more than 2 empty blocks in a row.

omerfirmak commented:
This is an extreme measure, could probably affect liveliness of our network. We are running our testnet with a much lower block interval, so things like this could work out when we switch to the 10 min block time.

I am wondering if what we should be doing is similar to the `confirmation` logic (ie. for a block to be confirmed and outputs in it to be spendable, there has to be certain number of blocks created after that height and longest chain is always the valid one in the case of a fork) that other chains implement. But that would be a big overhaul.

Nonetheless, we should still probably go with this. Let's rebase after #2670  and merge this too

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -455,6 +455,19 @@ extern(D):
+        {
+            auto last_block = this.ledger.getLastBlock();
+            auto signed = last_block.header.validators;
+            if (signed.setCount <= signed.count() / 2)
```
Nodes should also reject any nominations if they are missing signatures for the previous block, right?

hewison-chris replied:
Yes will add check.

hewison-chris replied:
Updated

hewison-chris committed:
Test.FutureEnrollment: Switch delayed node

This test has become a bit flaky recently but using node 0 as the
delayed node instead of node 5 seems to help.

hewison-chris committed:
Prefer candidate with more enrollments

hewison-chris committed:
CI integration test: allow up to 5 empty blocks
